### PR TITLE
Move query builder command handling to FunctionExecutionSupport interface

### PR DESCRIPTION
**Pull request by travisstebbins**
Move query builder command handling to `FunctionExecutionSupport` interface instead of to element-specific LSP grammar extensions (i.e., `ServiceLSPGrammarExtension`). This will prevent code duplication, because query builder commands will be used across various element types (services, functions, etc.) but they won't be dependent on the element type, so there's no need to duplicate the same switch/case handling of query builder commands in each LSP grammar extension. Instead, we can handle these commands in `FunctionExecutionSupport` with a new static `execute` command that the element-specific LSP grammar extensions can call.

### CODE REVIEW

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/AbstractLSPGrammarExtension.java [hunk 1]
```diff
@@ -253,6 +253,11 @@ protected void collectCommands(SectionState sectionState, PackageableElement ele
         this.commandsSupports.forEach(x -> x.collectCommands(sectionState, element, consumer));
     }
 
+    public Iterable<? extends LegendExecutionResult> defaultExecute(SectionState section, String entityPath, String commandId, Map<String, String> executableArgs, Map<String, Object> inputParameters)
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/AbstractLSPGrammarExtension.java:256]: Dont think we should create this method.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/connection/ConnectionLSPGrammarExtension.java [hunk 2]
```diff
@@ -92,7 +93,9 @@ protected void collectCommands(SectionState sectionState, PackageableElement ele
     @Override
     public Iterable<? extends LegendExecutionResult> execute(SectionState section, String entityPath, String commandId, Map<String, String> executableArgs, Map<String, Object> inputParams)
     {
-        return GENERATE_DB_COMMAND_ID.equals(commandId) ? generateDBFromConnection(section, entityPath) : super.execute(section, entityPath, commandId, executableArgs, Map.of());
+        return GENERATE_DB_COMMAND_ID.equals(commandId)
+               ? generateDBFromConnection(section, entityPath)
+               : FunctionExecutionSupport.execute(this, section, entityPath, commandId, executableArgs, inputParams);
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/connection/ConnectionLSPGrammarExtension.java:98]: Dont think we should change connection extension as this does not implement the function execution support interface.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java [hunk 3]
```diff
@@ -124,6 +121,45 @@ default String getExecutionKey(PackageableElement element, Map<String, Object> a
         return "";
     }
 
+    static Iterable<? extends LegendExecutionResult> execute(AbstractLSPGrammarExtension extension, SectionState section, String entityPath, String commandId, Map<String, String> executableArgs, Map<String, Object> inputParameters)
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java:124]: I would make extension of type FunctionExecutionSupport.  This APIs are for extensions that support function execution.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java [hunk 4]
```diff
@@ -124,6 +121,45 @@ default String getExecutionKey(PackageableElement element, Map<String, Object> a
...
+            {
+                return FunctionExecutionSupport.generateDatasetSpecifications(extension, section, entityPath, executableArgs, inputParameters);
+            }
+            case FunctionExecutionSupport.CHECK_DATASET_ENTITLEMENTS_ID:
+            {
+                return FunctionExecutionSupport.generateEntitlementReports(extension, section, entityPath, executableArgs, inputParameters);
+            }
+            default:
+            {
+                return extension.defaultExecute(section, entityPath, commandId, executableArgs, inputParameters);
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java:158]: Let's return null, and the caller then make the decision what to do.  Or, we can just throw with command not supported...

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java [hunk 5]
```diff
@@ -229,7 +265,7 @@ static Iterable<? extends LegendExecutionResult> executeFunction(FunctionExecuti
 
                 return Tuples.pair(executionPlan, planExecutionContext);
             });
-            executePlan(globalState, executionSupport, section.getDocumentState().getDocumentId(), section.getSectionNumber(), executionPlanAndContext.getOne(), executionPlanAndContext.getTwo(), entityPath, inputParameters, results);
+            executePlan(globalState, extension, section.getDocumentState().getDocumentId(), section.getSectionNumber(), executionPlanAndContext.getOne(), executionPlanAndContext.getTwo(), entityPath, inputParameters, results);
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java:268]: Lets keep this using executionSupport

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java [hunk 6]
```diff
@@ -238,10 +274,8 @@ static Iterable<? extends LegendExecutionResult> executeFunction(FunctionExecuti
         return results;
     }
 
-    static Iterable<? extends LegendExecutionResult> executeQuery(FunctionExecutionSupport executionSupport, SectionState section, String entityPath, Map<String, String> executableArgs, Map<String, Object> inputParameters)
+    static Iterable<? extends LegendExecutionResult> executeQuery(AbstractLSPGrammarExtension extension, SectionState section, String entityPath, Map<String, String> executableArgs, Map<String, Object> inputParameters)
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java:277]: Lets rollback this change, and all other similar changes.  These utilities should be for function executoin, and not all extensions support this.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/PureLSPGrammarExtension.java [hunk 7]
```diff
@@ -316,10 +316,10 @@ private Stream<Optional<LegendReferenceResolver>> toReferences(Constraint constr
     @Override
     public Iterable<? extends LegendExecutionResult> execute(SectionState section, String entityPath, String commandId, Map<String, String> executableArgs, Map<String, Object> inputParameters)
     {
-        return FunctionExecutionSupport.EXECUTE_COMMAND_ID.equals(commandId) ?
-                FunctionExecutionSupport.executeFunction(this, section, entityPath, inputParameters)
-                :
-                super.execute(section, entityPath, commandId, executableArgs, inputParameters);
+        return FunctionExecutionSupport.EXECUTE_COMMAND_ID.equals(commandId)
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/PureLSPGrammarExtension.java:319]: Wont the execute command also be handled by the function support?  I would expect this to just be

```return FunctionExecutionSupport.execute(this, section, entityPath, commandId, executableArgs, inputParameters);```

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/mapping/MappingLSPGrammarExtension.java [hunk 8]
```diff
@@ -188,7 +189,7 @@ public Iterable<? extends LegendExecutionResult> execute(SectionState section, S
             }
             default:
             {
-                return super.execute(section, entityPath, commandId, executableArgs, Map.of());
+                return FunctionExecutionSupport.execute(this, section, entityPath, commandId, executableArgs, inputParams);
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/mapping/MappingLSPGrammarExtension.java:192]: We dont need this on mapping extension.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/relational/RelationalLSPGrammarExtension.java [hunk 9]
```diff
@@ -205,7 +206,9 @@ protected void collectCommands(SectionState sectionState, PackageableElement ele
     @Override
     public Iterable<? extends LegendExecutionResult> execute(SectionState section, String entityPath, String commandId, Map<String, String> executableArgs, Map<String, Object> inputParams)
     {
-        return GENERATE_MODEL_MAPPING_COMMAND_ID.equals(commandId) ? generateModelsFromDatabaseSpecification(section, entityPath) : super.execute(section, entityPath, commandId, executableArgs, Map.of());
+        return GENERATE_MODEL_MAPPING_COMMAND_ID.equals(commandId)
+               ? generateModelsFromDatabaseSpecification(section, entityPath)
+               : FunctionExecutionSupport.execute(this, section, entityPath, commandId, executableArgs, inputParams);
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/relational/RelationalLSPGrammarExtension.java:211]: Similar to other comments, we dont need this on relational.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/service/ServiceLSPGrammarExtension.java [hunk 10]
```diff
@@ -305,37 +305,9 @@ public Iterable<? extends LegendExecutionResult> execute(SectionState section, S
...
-                return FunctionExecutionSupport.generateDatasetSpecifications(this, section, entityPath, executableArgs, inputParams);
-            }
-            case FunctionExecutionSupport.CHECK_DATASET_ENTITLEMENTS_ID:
-            {
-                return FunctionExecutionSupport.generateEntitlementReports(this, section, entityPath, executableArgs, inputParams);
-            }
             default:
             {
-                return super.execute(section, entityPath, commandId, executableArgs, Map.of());
+                return FunctionExecutionSupport.execute(this, section, entityPath, commandId, executableArgs, inputParams);
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/service/ServiceLSPGrammarExtension.java:310]: We should handle execute command here too, and remove it from the switch.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/AbstractLSPGrammarExtension.java (hunk 1 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/AbstractLSPGrammarExtension.java]: Removed it

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/connection/ConnectionLSPGrammarExtension.java (hunk 2 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/connection/ConnectionLSPGrammarExtension.java]: Makes sense. Reverted the change to this file.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java (hunk 3 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java]: Done

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java (hunk 4 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java]: Done (throwing an error now)

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java (hunk 5 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java]: Done

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java (hunk 6 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/FunctionExecutionSupport.java]: Done

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/PureLSPGrammarExtension.java (hunk 7 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/core/PureLSPGrammarExtension.java]: Done

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/mapping/MappingLSPGrammarExtension.java (hunk 8 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/mapping/MappingLSPGrammarExtension.java]: Makes sense. Reverted the change to this file.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/relational/RelationalLSPGrammarExtension.java (hunk 9 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/relational/RelationalLSPGrammarExtension.java]: Makes sense. Reverted the change to this file.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/service/ServiceLSPGrammarExtension.java (hunk 10 above)
Comment (travisstebbins) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/service/ServiceLSPGrammarExtension.java:306]: Done

**Review by rafaelbey — APPROVED** [review]

**Pull request merged**